### The aim of this project is to clean the data and analyse the included used car listings

In [1]:
import pandas as pd
import numpy as np
import datetime as dt

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos.info()
autos.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

#### Observations:
1) The dataframe has maximum of 50000 row, there are columns with empty cells
<br>
2) There are capital letters used in header row
<br>

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns =['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

#### Changes:
Header has been changed from camelcase to snakecase to facilitate better column identification

In [6]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

#### Observations:
1) registration_year - There is a cell with 9999 which is invalid
<br>
2) seller, offer_type, unrepaired_damage, nr_of_pictures - Can be dropped as cell are mainly unique values
<br>
3) price - in text form, should be in numeric data form
<br>
4) odometer - in text form, should be in numeric data form

In [7]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")

autos["odometer_km"] = autos["odometer_km"].str.replace("km","")
autos["odometer_km"] = autos["odometer_km"].str.replace(",","")

In [8]:
autos["price"] = autos["price"].astype(float)
autos["odometer_km"] = autos["odometer_km"].astype(float)

#### Changes:
Price and odometer_km column has been changed to numeric data

## Exploring odometer_km column

In [9]:
autos["odometer_km"].unique().shape

(13,)

In [10]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [11]:
autos["odometer_km"].value_counts().sort_index()

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

#### Observation:
There is no issue with the odometer_km column

## Exploring price column

In [12]:
autos["price"].unique().shape

(2357,)

In [13]:
autos["price"].describe().round()

count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price, dtype: float64

In [14]:
autos["price"].value_counts().sort_index(ascending = False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

#### Observation:
The high outlier is obvious and can be removed without exploration

In [15]:
autos["price"].value_counts().sort_index(ascending = True).head(15)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: price, dtype: int64

#### Observation:
The low outlier has a larger number of vehicle, a scatterplot is recommended but instead will be filtering listing with price below 1000

In [16]:
autos = autos[(autos["price"] > 999) & (autos["price"] < 350001)]

## Converting datetime data

In [17]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [18]:
autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index().round(3)

2016-03-05    0.026
2016-03-06    0.014
2016-03-07    0.035
2016-03-08    0.033
2016-03-09    0.032
2016-03-10    0.033
2016-03-11    0.033
2016-03-12    0.037
2016-03-13    0.016
2016-03-14    0.037
2016-03-15    0.034
2016-03-16    0.029
2016-03-17    0.030
2016-03-18    0.013
2016-03-19    0.035
2016-03-20    0.038
2016-03-21    0.037
2016-03-22    0.033
2016-03-23    0.032
2016-03-24    0.029
2016-03-25    0.031
2016-03-26    0.033
2016-03-27    0.031
2016-03-28    0.035
2016-03-29    0.034
2016-03-30    0.033
2016-03-31    0.031
2016-04-01    0.035
2016-04-02    0.036
2016-04-03    0.039
2016-04-04    0.037
2016-04-05    0.013
2016-04-06    0.003
2016-04-07    0.002
Name: date_crawled, dtype: float64

#### Observation:
Data for date_crawled is evenly distributed

In [19]:
autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index().round(5)

2015-06-11    0.00003
2015-08-10    0.00003
2015-09-09    0.00003
2015-11-10    0.00003
2015-12-30    0.00003
2016-01-03    0.00003
2016-01-07    0.00003
2016-01-10    0.00005
2016-01-13    0.00003
2016-01-14    0.00003
2016-01-16    0.00003
2016-01-22    0.00003
2016-01-27    0.00008
2016-01-29    0.00003
2016-02-01    0.00003
2016-02-02    0.00005
2016-02-05    0.00005
2016-02-07    0.00003
2016-02-09    0.00003
2016-02-11    0.00003
2016-02-12    0.00005
2016-02-14    0.00005
2016-02-16    0.00003
2016-02-17    0.00003
2016-02-18    0.00005
2016-02-19    0.00008
2016-02-20    0.00003
2016-02-21    0.00005
2016-02-22    0.00003
2016-02-23    0.00010
               ...   
2016-03-09    0.03265
2016-03-10    0.03298
2016-03-11    0.03306
2016-03-12    0.03713
2016-03-13    0.01766
2016-03-14    0.03498
2016-03-15    0.03345
2016-03-16    0.02964
2016-03-17    0.03016
2016-03-18    0.01328
2016-03-19    0.03407
2016-03-20    0.03826
2016-03-21    0.03757
2016-03-22    0.03231
2016-03-23

#### Observation
Date for ad_created increases overtime

In [20]:
autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index().round(3)

2016-03-05    0.001
2016-03-06    0.004
2016-03-07    0.005
2016-03-08    0.006
2016-03-09    0.009
2016-03-10    0.010
2016-03-11    0.012
2016-03-12    0.022
2016-03-13    0.008
2016-03-14    0.012
2016-03-15    0.015
2016-03-16    0.015
2016-03-17    0.026
2016-03-18    0.007
2016-03-19    0.015
2016-03-20    0.020
2016-03-21    0.020
2016-03-22    0.021
2016-03-23    0.018
2016-03-24    0.019
2016-03-25    0.018
2016-03-26    0.016
2016-03-27    0.014
2016-03-28    0.019
2016-03-29    0.021
2016-03-30    0.023
2016-03-31    0.023
2016-04-01    0.023
2016-04-02    0.025
2016-04-03    0.024
2016-04-04    0.023
2016-04-05    0.131
2016-04-06    0.235
2016-04-07    0.140
Name: last_seen, dtype: float64

#### Observation:
Data for last_seen is evenly distributed

In [21]:
autos["registration_year"].describe()

count    38626.000000
mean      2005.679801
std         86.685138
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

#### Observation:
Min of 1000 and Max of 9999 is invalid, need to remove outliers

In [22]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       3
1959       6
1960      17
1961       6
1962       4
1963       8
1964       9
1965      17
1966      21
1967      25
        ... 
1999    1724
2000    2001
2001    2065
2002    2135
2003    2477
2004    2608
2005    2785
2006    2651
2007    2259
2008    2204
2009    2076
2010    1584
2011    1617
2012    1306
2013     795
2014     649
2015     361
2016     651
2017    1008
2018     398
2019       1
2800       1
4100       1
4500       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       2
Name: registration_year, Length: 91, dtype: int64

#### Observation:
Year 1000 and 1001 is invalid as cars is not available yet
Years that are higher than 2019 is invalid as it is later than current year

In [23]:
autos = autos[(autos["registration_year"] > 1926) & (autos["registration_year"] < 2020)]

In [24]:
autos["registration_year"].value_counts(normalize = True)

2005    0.072126
2006    0.068656
2004    0.067542
2003    0.064149
2007    0.058504
2008    0.057079
2002    0.055292
2009    0.053764
2001    0.053479
2000    0.051822
1999    0.044648
2011    0.041877
2010    0.041022
2012    0.033823
1998    0.033227
2017    0.026105
2013    0.020589
1997    0.020382
2016    0.016860
2014    0.016808
1996    0.014037
1995    0.011473
2018    0.010307
2015    0.009349
1994    0.007018
1993    0.005646
1992    0.005646
1991    0.005361
1990    0.004895
1989    0.003185
          ...   
1977    0.000518
1975    0.000466
1969    0.000466
1965    0.000440
1960    0.000440
1964    0.000233
1963    0.000207
1959    0.000155
1961    0.000155
1937    0.000104
1962    0.000104
1956    0.000104
1958    0.000078
1954    0.000052
1934    0.000052
1941    0.000052
1951    0.000052
1957    0.000052
1955    0.000052
1929    0.000026
1953    0.000026
1943    0.000026
2019    0.000026
1939    0.000026
1950    0.000026
1931    0.000026
1938    0.000026
1948    0.0000

#### Observation:
Listing for cars increase over the years

## Aggregating the Brand Column

In [25]:
autos["brand"].value_counts(normalize = True).mul(100).round(1).astype(str) + '%'

volkswagen        21.3%
bmw               12.4%
mercedes_benz     11.1%
audi               9.7%
opel               9.0%
ford               5.9%
renault            3.8%
peugeot            2.8%
fiat               2.1%
skoda              1.9%
seat               1.8%
smart              1.7%
toyota             1.5%
mazda              1.4%
citroen            1.4%
nissan             1.4%
hyundai            1.1%
mini               1.1%
sonstige_autos     1.0%
volvo              0.9%
kia                0.8%
honda              0.7%
porsche            0.7%
mitsubishi         0.7%
chevrolet          0.7%
alfa_romeo         0.6%
suzuki             0.6%
dacia              0.3%
chrysler           0.3%
jeep               0.3%
land_rover         0.3%
jaguar             0.2%
daihatsu           0.2%
subaru             0.2%
saab               0.1%
daewoo             0.1%
trabant            0.1%
rover              0.1%
lancia             0.1%
lada               0.1%
Name: brand, dtype: object

#### Observation:
Brands that are >5% will be selected, which adds up to six brands
<br>
Brands: Volkwagen, BMW, Mercedes Benz, Audi, Opel, Ford

In [36]:
autos_brand = autos[autos.brand.isin(["volkswagen", "bmw", "mercedes_benz", "audi", "opel", "ford"])]
autos_brand.groupby("brand").price.mean()

brand
audi             10200.053490
bmw               9055.438494
ford              5292.868549
mercedes_benz     9215.087933
opel              4177.930796
volkswagen        6556.728822
Name: price, dtype: float64

In [39]:
autos_brand.groupby("brand").odometer_km.mean()

brand
audi             127897.833645
bmw              132118.200837
ford             119841.199824
mercedes_benz    130164.873714
opel             124124.855825
volkswagen       126055.866602
Name: odometer_km, dtype: float64

In [27]:
autos_brand.head(10)

date_crawled                                               name  \
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
20  2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
24  2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
29  2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31  2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   

    seller offer_type    price   abtest vehicle_type  registration_year  \
1   privat    Angebot   8500.0  control    limousine               1997   
2   privat    Angebot   8990.0     test    limousine               2009   
4   privat    Angebot   1350.0     test        kombi               2003   
7   privat    Angebot   1990.0  control    limousine               1998   
13  privat    Angebot   1350.0  control    limousine               1999   
17  privat    Angebot   5500.0     test        coupe               1990   
20  privat    Angebot   3500.0     test        kombi               2003   
24  privat    Angebot  48500.0  control    limousine               2014   
29  privat    Angebot   4999.0     test        kombi               2004   
31  privat    Angebot   2850.0     test        kombi               2002   

      gearbox  power_ps     model  odometer_km  registration_month fuel_type  \
1   automatik       286       7er     150000.0                   6    benzin   
2     manuell       102      golf      70000.0                   7    benzin   
4     manuell         0     focus     150000.0                   7    benzin   
7     manuell        90      golf     150000.0                  12    diesel   
13    manuell       101        a3     150000.0                  11    benzin   
17    manuell       205  scirocco     150000.0                   6    benzin   
20    manuell       131        a4     150000.0                   5    diesel   
24  automatik       306       5er      30000.0                  12    benzin   
29  automatik       204  e_klasse     150000.0                  10    diesel   
31  automatik       218  c_klasse     150000.0                   7    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
7      volkswagen              nein  2016-03-16 00:00:00               0   
13           audi              nein  2016-03-23 00:00:00               0   
17     volkswagen              nein  2016-03-29 00:00:00               0   
20           audi               NaN  2016-03-05 00:00:00               0   
24            bmw              nein  2016-04-03 00:00:00               0   
29  mercedes_benz              nein  2016-04-02 00:00:00               0   
31  mercedes_benz              nein  2016-03-14 00:00:00               0   

    postal_code            last_seen  
1         71034  2016-04-06 14:45:08  
2         35394  2016-04-06 20:15:37  
4         39218  2016-04-01 14:38:50  
7         53474  2016-04-07 03:17:32  
13        12043  2016-04-01 14:17:13  
17        74821  2016-04-05 20:46:26  
20        53913  2016-03-07 05:46:46  
24        22547  2016-04-07 13:16:50  
29        47638  2016-04-02 12:45:44  
31        41065  2016-03-16 07:19:04

In [35]:
# Platform Answer

brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')


{'audi': 10200,
 'bmw': 9055,
 'ford': 5292,
 'mercedes_benz': 9215,
 'opel': 4177,
 'volkswagen': 6556}

#### Conclusion:
In this project data cleaning and analysis is performed
<br>

1) Header rows is formated to snake case for easier label calling
<br>
2) Outlier in the price column is removed
<br>
3) Invalid years in the registration year is removed
<br>
4) date_crawled, ad_created and last_seen column converted to datetime values
<br>
5) Identified the brands with most listings and created a new dataframe with only those brands
<br>
6) Calcuated the mean prices and mean mileage of brands with most listings
<br>

Audi, BMW and Mercedes Benz is relatively more expensive that Ford, Opel and Volkswagen
<br>
Mileage is similar across all brands. Therefore the price of the more expensive brand does not offer value interms of mileage